download_generic.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"github.com/hashicorp/go-getter"
+	"github.com/sirupsen/logrus"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// Calculates the md5sum of a local file
+func md5sum(path string) (string, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	hash := md5.New()
+	_, err = io.Copy(hash, file)
+	if err != nil {
+		return "", err
+	}
+
+	byteHash := hash.Sum(nil)[:16]
+	return hex.EncodeToString(byteHash), nil
+}
+
+func idempotentGenericFileDownload(resource, path string) error {
+	logrus.Debugf("Starting idempotent download of %s", resource)
+	hashResource := resource + ".md5"
+
+	currentChecksum, err := md5sum(path)
+	if os.IsNotExist(err) {
+		currentChecksum = ""
+	} else if err != nil {
+		return err
+	}
+
+	dir, err := ioutil.TempDir("", "*")
+	if err != nil {
+		logrus.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+	hashFile := filepath.Join(dir, "md5Hash")
+
+	var remoteChecksum string
+	err = getter.GetFile(hashResource, hashFile)
+	if err != nil {
+		logrus.Debugf("MD5 sum not found at: %s", hashResource)
+		remoteChecksum = ""
+	} else {
+		logrus.Debugf("Found MD5 sum at: %s", hashResource)
+		content, err := ioutil.ReadFile(hashFile)
+		if err != nil {
+			logrus.Debug("Error reading remote checksum")
+			return err
+		}
+		remoteChecksum = string(content)
+	}
+
+	if currentChecksum != "" && remoteChecksum != "" {
+		logrus.Debugf("Local checksum:  %s", currentChecksum)
+		logrus.Debugf("Remote checksum: %s", remoteChecksum)
+		if string(remoteChecksum) == currentChecksum {
+			logrus.Debug("Local and remote checksums match, skipping file download")
+			return nil
+		}
+	}
+
+	if remoteChecksum != "" {
+		logrus.Debugf("Downloading file: %s", resource)
+		return getter.GetFile(path, fmt.Sprintf("%s?checksum=md5:%s", resource, remoteChecksum))
```

Comment:
Ah, took me a minute. So we verified that there is a new checksum and that it isn't the same as the old checksum, so we are just validating that this new file matches the new checksum. Makes sense.